File: tests/fixtures/java-library/src/main/java/library/extensions/Results.java
package library.extensions;

import library.models.Book;

/** Extension: sealed interface hierarchy. */
public sealed interface SearchResult permits SearchResult.Found, SearchResult.NotFound, SearchResult.Error {

    /** Found a match with a relevance score. */
    record Found(Book book, double score) implements SearchResult {}

    /** No results for the query. */
    record NotFound(String query) implements SearchResult {}

    /** Search error with message. */
    record Error(String message, int code) implements SearchResult {}

    /** Check if this result contains a match — extension: pattern matching. */
    default boolean isMatch() {
        return this instanceof Found;
    }
}

File: tests/fixtures/kotlin-library/src/main/kotlin/library/extensions/Results.kt
package library.extensions

import library.models.Book

/** Extension: sealed class with data class, object, and class subclasses. */
sealed class SearchResult {
    /** Found a match with a relevance score. */
    data class Found(val book: Book, val score: Double) : SearchResult()

    /** No results for the query. */
    object NotFound : SearchResult()

    /** Search error with message. */
    data class Error(val message: String, val code: Int) : SearchResult()

    /** Check if this result contains a match. */
    fun isMatch(): Boolean = this is Found
}

/** Extension: object declaration (singleton). */
object BookRegistry {
    private val books = mutableMapOf<String, Book>()

    fun register(book: Book) {
        books[book.isbn] = book
    }

    fun lookup(isbn: String): Book? = books[isbn]
}
